test/Conversion/TorchToSCF/basic.mlir

```diff
@@ -64,3 +64,37 @@ func @aten.prim.if$nested(%arg0: !torch.bool, %arg1: !torch.bool) -> !torch.int
   }
   return %0 : !torch.int
 }
+
+  // CHECK-LABEL: func @torch.prim.loop$while
+  // CHECK-SAME:  (%[[ARG0:.*]]: !torch.int) -> !torch.float {
+  // CHECK:         %[[TORCH_FLOAT_VAL:.*]] = torch.constant.float
+  // CHECK-NEXT:    %[[FLOAT_VAL:.*]] = torch_c.to_f64 %[[TORCH_FLOAT_VAL]]
+  // CHECK-NEXT:    %[[MAX_TRIP_COUNT:.*]] = torch.constant.int 9223372036854775807
+  // CHECK-NEXT:    %[[LOOP:.*]] = scf.while
+  // CHECK-SAME:    (%[[LOOP_ARG:.*]] = %[[FLOAT_VAL]]) : (f64) -> f64 {
+  // CHECK-NEXT:      %[[TORCH_LOOP_ARG:.*]] = torch_c.from_f64 %[[LOOP_ARG]]
+  // CHECK-NEXT:      %[[TORCH_CONDITION:.*]] = torch.aten.lt.float_int %[[TORCH_LOOP_ARG]], %[[ARG0]]
+  // CHECK-NEXT:      %[[CONDITION:.*]] = torch_c.to_i1 %[[TORCH_CONDITION]]
+  // CHECK-NEXT:      %[[CONDITION_:.*]] = torch_c.to_i1 %[[TORCH_CONDITION]]
+  // CHECK-NEXT:      scf.condition(%[[CONDITION_]]) %[[LOOP_ARG]]
+  // CHECK-NEXT:    } do {
+  // CHECK-NEXT:    ^bb0(%[[BLOCK_ARG:.*]]: f64):
+  // CHECK-NEXT:      %[[TORCH_BLOCK_ARG:.*]] = torch_c.from_f64 %[[BLOCK_ARG]]
+  // CHECK-NEXT:      %[[TORCH_VAL:.*]] = torch.aten.mul.float %[[TORCH_BLOCK_ARG]], %[[TORCH_BLOCK_ARG]]
```

Comment:
why is this op in the "after" region? I think you should be able to do the lowering by just putting everything in the "before" region and leaving the "after" region empty.

Reply:
The before region of the `scf.while` contains the condition of the while loop while the after region has the body of the loop. That is why I coded it like that.

Reply:
Even the lowering for `mhlo.while` to `scf.while` followed the same, which I took as a reference. https://github.com/tensorflow/mlir-hlo/blob/078a98ec5b7127fd46c050c8bcb81e6e97ad0ecb/lib/Dialect/mhlo/transforms/legalize_control_flow.cc#L78

Reply:
I don't think there is any such restriction for scf.while. You should be able to put the whole computation in the "before" region.

Reply:
Or another alternative is to calculate the condition in the after region and the pass it through the iterargs into the before region -- I think you might need to do that to handle the initial condition cleanly.

Reply:
mhlo.while structurally different from a while-like PrimLoopOp though. A while-like PrimLoopOp is actually do-while-like with a peeled initial check for the condition.

Reply:
Yeah, I think the natural lowering is to put the body of the PrimLoopOp into the "after" region of scf.while, and then pass the condition as an iterArg back to the "before" region. Then the PrimLoopOp initialCondition is handled by passing it as an iterArg to scf.while.

Reply:
It's unfortunate that PrimLoopOp is defined this way, because it forces the frontend to peel the condition, which make it hard to inline large conditions. Here is an example:

```
def f(x):
    while x.sum() > 0:
        x = x * x
    return x
```

```
graph(%x.1 : Tensor):
  %22 : NoneType = prim::Constant()
  %1 : int = prim::Constant[value=9223372036854775807]() # /tmp/repro.py:4:4
  %5 : int = prim::Constant[value=0]() # /tmp/repro.py:4:20
   // Notice the aten::sum/gt/Bool is replicated here...
  %23 : Tensor = aten::sum(%x.1, %22) # /tmp/repro.py:4:10
  %24 : Tensor = aten::gt(%23, %5) # /tmp/repro.py:4:10
  %26 : bool = aten::Bool(%24) # /tmp/repro.py:4:10
  %x : Tensor = prim::Loop(%1, %26, %x.1) # /tmp/repro.py:4:4
    block0(%9 : int, %x.15 : Tensor):
      %x.9 : Tensor = aten::mul(%x.15, %x.15) # /tmp/repro.py:5:12
     // ... and replicated here
      %4 : Tensor = aten::sum(%x.9, %22) # /tmp/repro.py:4:10
      %6 : Tensor = aten::gt(%4, %5) # /tmp/repro.py:4:10
      %8 : bool = aten::Bool(%6) # /tmp/repro.py:4:10
      -> (%8, %x.9)
  return (%x)
```

Reply:
Thanks @silvasean for your insights. Will make necessary changes in the code.